akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -56,6 +57,37 @@ private[akka] final case class Collect[In, Out](decider: Supervision.Decider)(pf
   override def decide(t: Throwable): Supervision.Directive = decider(t)
 }
 
+private[akka] final case class Recover[T](pf: PartialFunction[Throwable, T]) extends PushPullStage[T, T] {
+  import Collect.NotApplied
+  var recovered: Option[T] = None
+
+  override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
+    ctx.push(elem)
+  }
+
+  override def onPull(ctx: Context[T]): SyncDirective =
+    if (recovered.isDefined) {
+      val value = recovered.get
+      recovered = None
+      if (ctx.isFinishing) ctx.pushAndFinish(value) else ctx.push(value)
```

Comment:
it is always in finishing, since recovered is only set to False in onUpstreamFailure